### ISSUE TITLE
[Bug]: Bad performance / Very high CPU load when using INSERT DO UPDATE ON CONFLICT on compressed chunks

### ISSUE THREAD

**Issue by mblsf** [timescale/timescaledb#5801]
### What type of bug is this?

Performance issue

### What subsystems and features are affected?

Compression

### What happened?

When using the new feature INSERT DO UPDATE ON CONFLICT on compressed chunks that contain a lot of data. the performance is really bad (Processing 100 records takes 4 seconds instead of 40 ms) and the CPU load on postgres is always very high during this operation.

### TimescaleDB version affected

2.11.0

### PostgreSQL version used

14.8

### What operating system did you use?

Ubuntu 22.04 x64

### What installation method did you use?

Docker

### What platform did you run on?

On prem/Self-hosted

### How can we reproduce the bug?

```bash
Perform INSERT DO UPDATE ON CONFLICT on compressed chunks that contain a lot of data
```

**Comment by lkshminarayanan** [comment]
Hi @mblsf,

Thank you for filing the issue!
I tested INSERT DO UPDATE ON CONFLICT on a hypertable that had around 50M rows.
Without compression, the statement took around 23ms and with compression it took around 380ms.
This increase in time is expected due to the overhead of compression.

But I was not able to observe the level of performance degradation that you mention in the bug.
It could be the case that this degradation occurs only under certain conditions.
So, can you provide more information about your setup where you faced this issue?
Hypertable definition, number of rows in the hypertable, the complete query with which you faced the issue - anything would be helpful.

**Comment by mblsf** [comment]
Forget to mention that we are using batch inserts. The values to insert can be for different chunks. For the definition of hypertable definition I need to wait for a colleague to return. Query locks like this:

INSERT INTO values(key, ts_utc, value, state, id) VALUES
     ('key1', '2023-06-19 16:18:38.580076', 'value1', 'new', get_id()),
     ('key2', '2023-06-19 16:18:38.580324', 'value2', 'changed', get_id()),
    ....
     ('key100', '2023-06-19 16:18:38.674424', 'value100', 'deleted', get_id())
ON CONFLICT (key, ts_utc) DO UPDATE SET 
     value = EXCLUDED.value,
     state = EXCLUDED.state;
     
get_id() is used to generate a unique key

**Comment by sb230132** [comment]
Hi @mblsf 
If the INSERT is going to affect all segments in compressed chunk, then this INSERT operation will end up decompressing everything in compressed chunk which will be time consuming.

Prior 2.11.0 INSERT on compressed chunk will compress only the new row and insert into compressed chunk. From 2.11.0 
we extract (decompress) all those segments which will be affected and then do the INSERT operation. Thus you see a performance hit.

Anyways, we need details on hypertable, and compression settings like what are the segmentby columns to understand better on the issue.

**Comment by mblsf** [comment]
Here is the requested information. Hope it's complete. If not, please reply.

CREATE TABLE IF NOT EXISTS warp_tsa_core.signal_raw_values
(
     signal_key VARCHAR(50) NOT NULL ,
     raw_ts_utc TIMESTAMP NOT NULL ,
     raw_dupl_id INTEGER NOT NULL DEFAULT 0,
     raw_value_num FLOAT8,
     raw_value_string VARCHAR(1000),
     raw_state_name VARCHAR(1000),
     raw_good_flag BOOLEAN,
     load_id INTEGER NULL,
     CONSTRAINT signal_raw_values_key PRIMARY KEY
     ( signal_key, raw_ts_utc ,raw_dupl_id )
);

-- convert table to hypertable
SELECT warp_tsa_core.create_hypertable
       (relation => 'warp_tsa_core.signal_raw_values'
       ,time_column_name =>'raw_ts_utc'
       ,chunk_time_interval => INTERVAL '1 day'
       ,partitioning_column => 'signal_key'
       ,number_partitions => 1              -- 2 when using 2 data nodes; 4 when using 4 data nodes
       ,if_not_exists => TRUE
       ,migrate_data => TRUE
       );

-- Data older than 400 days in table signal_raw_values should be deleted automatically by timescaledb
SELECT warp_tsa_core.add_retention_policy('warp_tsa_core.signal_raw_values', INTERVAL '400 days');

-- Activates compression on signal_raw_values for data older than 30 days
ALTER TABLE warp_tsa_core.signal_raw_values SET (
  timescaledb.compress,
  timescaledb.compress_orderby = 'raw_ts_utc,raw_dupl_id',
  timescaledb.compress_segmentby = 'signal_key'
);

-- Set compress policy to automatic compress chunks older than 30 days
SELECT warp_tsa_core.add_compression_policy(
  'warp_tsa_core.signal_raw_values' -- hypertable
 ,INTERVAL '30d'          -- after interval compress
 ,true                -- no error (only warning) if compression policy already exists
 );

**Comment by RafiaSabih** [comment]
@mblsf Please provide the queries you are facing issues with also.

**Comment by S-imo-n** [comment]
Any updates yet? Do you need more information?

**Comment by mblsf** [comment]
@lkshminarayanan, @sb230132, @RafiaSabih Can someone of you please have a look at this issue again? I think we've recently provided all necessary information.

**Comment by mkindahl** [comment]
@mblsf I do not think that a large chunk size is is the reason for bad performance. Could you run the following queries to get some statistics on the columns?

```sql
SELECT schemaname, tablename, attname, n_distinct
  FROM pg_stats
 WHERE schemaname = 'warp_tsa_core'
   AND tablename = 'signal_raw_values'
   AND inherited;
```

**Comment by sb230132** [comment]
Here is the issue which should fix the performance in INSERTS ... ON CONFLICT queries on compressed chunks.
https://github.com/timescale/timescaledb/issues/6063

**Comment by mblsf** [comment]
@mkindahl Here are the results for your query, if still relevant:

"warp_tsa_core"	"signal_raw_values"	"signal_key"	22309
"warp_tsa_core"	"signal_raw_values"	"raw_ts_utc"	1.261464e+06
"warp_tsa_core"	"signal_raw_values"	"raw_dupl_id"	1
"warp_tsa_core"	"signal_raw_values"	"raw_value_num"	143824
"warp_tsa_core"	"signal_raw_values"	"raw_value_string"	708
"warp_tsa_core"	"signal_raw_values"	"raw_state_name"	113
"warp_tsa_core"	"signal_raw_values"	"raw_good_flag"	2
"warp_tsa_core"	"signal_raw_values"	"load_id"	14295

**Comment by mkindahl** [comment]
> @mkindahl Here are the results for your query, if still relevant:
> 
> "warp_tsa_core" "signal_raw_values" "signal_key" 22309 "warp_tsa_core" "signal_raw_values" "raw_ts_utc" 1.261464e+06 "warp_tsa_core" "signal_raw_values" "raw_dupl_id" 1 "warp_tsa_core" "signal_raw_values" "raw_value_num" 143824 "warp_tsa_core" "signal_raw_values" "raw_value_string" 708 "warp_tsa_core" "signal_raw_values" "raw_state_name" 113 "warp_tsa_core" "signal_raw_values" "raw_good_flag" 2 "warp_tsa_core" "signal_raw_values" "load_id" 14295

Thank you. It seems you picked a good segment-by key, and I just wanted to eliminate that you had picked a bad segment-by key.

**Comment by mblsf** [comment]
@mkindahl You're welcome. Please let me know when there are any updates related to this issue.

**Comment by S-imo-n** [comment]
Hi @mkindahl any updates?

**Comment by aarondglover** [comment]
Using 2.13 on PG15 in AIVEN/MST.

I have not measured the speed difference of INSERT ON CONFLICT etc but I what I have observed is that UNIQUE INDEXES are not observed. It allows for duplicate rows to be inserted as per my recent comments in #6302

Re comments from @sb230132 :

> Prior 2.11.0 INSERT on compressed chunk will compress only the new row and insert into compressed chunk. From 2.11.0
> we extract (decompress) all those segments which will be affected and then do the INSERT operation. Thus you see a performance hit.

I'm not sure the decompression is occurring prior to the insertion into a compressed chunk - I think this because the unique index is not being triggered and I am ending up with duplicate rows.  One row in the compressed chunk, and one duplicate in the uncompressed part of the chunk. Once the duplicate row is in there, decompression of chunk into the uncompressed chunk then fails because the UNIQUE INDEX borks on the duplicate row.

**Comment by aarondglover** [comment]
Is DML on compressed chunks fundamentally broken??? It would seems so to me 😢

**Comment by svenklemm** [comment]
There is a bug preventing this from happening with unique expression indexes, for non-expression indexes it should work in 2.13.
See https://github.com/timescale/timescaledb/pull/6454

**Comment by antekresic** [comment]
INSERT ON CONFLICT DO UPDATE received optimizations recently that should have helped with this. They are available with version 2.17

Have you had the chance to try it out?

**Comment by S-imo-n** [comment]
It worked, thanks :)

**Comment by akuzm** [comment]
Some problems mentioned in this issue were fixed, please create a new one if there are any other remaining problems.
